src/envs/pybullet_blocks.py

```diff
@@ -0,0 +1,728 @@
+                                     baseVisualShapeIndex=visual_id,
+                                     basePosition=pose,
+                                     baseOrientation=self._block_orientation,
+                                     physicsClientId=self._physics_client_id)
+        p.changeDynamics(block_id,
+                         -1,
+                         lateralFriction=self._block_friction,
+                         physicsClientId=self._physics_client_id)
+
+        return block_id
+
+    def render(
+            self,
+            action: Optional[Action] = None,
+            caption: Optional[str] = None) -> List[Image]:  # pragma: no cover
+        # Skip test coverage because GUI is too expensive to use in unit tests
+        # and cannot be used in headless mode.
+        del caption  # unused
+
+        if not CFG.pybullet_use_gui:
+            raise Exception(
+                "Rendering only works with GUI on. See "
+                "https://github.com/bulletphysics/bullet3/issues/1157")
+
+        view_matrix = p.computeViewMatrixFromYawPitchRoll(
+            cameraTargetPosition=self._camera_target,
+            distance=self._camera_distance,
+            yaw=self._camera_yaw,
+            pitch=self._camera_pitch,
+            roll=0,
+            upAxisIndex=2,
+            physicsClientId=self._physics_client_id)
+
+        width = CFG.pybullet_camera_width
+        height = CFG.pybullet_camera_height
+
+        proj_matrix = p.computeProjectionMatrixFOV(
+            fov=60,
+            aspect=float(width / height),
+            nearVal=0.1,
+            farVal=100.0,
+            physicsClientId=self._physics_client_id)
+
+        (_, _, px, _,
+         _) = p.getCameraImage(width=width,
+                               height=height,
+                               viewMatrix=view_matrix,
+                               projectionMatrix=proj_matrix,
+                               renderer=p.ER_BULLET_HARDWARE_OPENGL,
+                               physicsClientId=self._physics_client_id)
+
+        rgb_array = np.array(px)
+        rgb_array = rgb_array[:, :, :3]
+        return [rgb_array]
+
+    def step(self, action: Action) -> State:
+        ee_delta, finger_action = action.arr[:3], action.arr[3]
+
+        ee_link_state = p.getLinkState(self._fetch_id,
+                                       self._ee_id,
+                                       physicsClientId=self._physics_client_id)
+        current_position = ee_link_state[4]
+        target_position = np.add(current_position, ee_delta).tolist()
+
+        # We assume that the robot is already close enough to the target
+        # position that IK will succeed with one call, so validate is False.
+        # Furthermore, updating the state of the robot during simulation, which
+        # validate=True would do, is risky and discouraged by PyBullet.
+        joint_values = inverse_kinematics(
```

Comment:
I would rather leave it like this because if we copy and paste this code into another repository some day (like I did in this instance), it's more obvious that we just need to change the default in the kwarg, vs having to introspect the function for CFGs.